contracts/access_control/lib.rs

```diff
@@ -0,0 +1,251 @@
+        }
+
+        /// Gives a role to an account
+        ///
+        /// Can only be called by an account with an admin role on this contract
+        #[ink(message, selector = 1)]
+        pub fn grant_role(&mut self, account: AccountId, role: Role) -> Result<()> {
+            let key = (account, role);
+            if !self.priviledges.contains(key) {
+                let caller = self.env().caller();
+                let this = self.env().account_id();
+                self.check_role(caller, Role::Admin(this))?;
+                self.priviledges.insert(key, &());
+
+                let event = Event::RoleGranted(RoleGranted {
+                    by: caller,
+                    to: account,
+                    role,
+                });
+                Self::emit_event(self.env(), event);
+            }
+
+            Ok(())
+        }
+
+        /// Revokes a role from an account
+        ///
+        /// Can only be called by an admin role on this contract
+        #[ink(message, selector = 2)]
+        pub fn revoke_role(&mut self, account: AccountId, role: Role) -> Result<()> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            self.check_role(caller, Role::Admin(this))?;
+            self.priviledges.remove((account, role));
+
+            let event = Event::RoleRevoked(RoleRevoked {
+                by: caller,
+                from: account,
+                role,
+            });
+            Self::emit_event(self.env(), event);
+
+            Ok(())
+        }
+
+        /// Returns true if account has a role
+        #[ink(message, selector = 3)]
+        pub fn has_role(&self, account: AccountId, role: Role) -> bool {
+            self.priviledges.get((account, role)).is_some()
+        }
+
+        /// Terminates the contract.
+        ///
+        /// can only be called by the contract owner
+        #[ink(message, selector = 4)]
+        pub fn terminate(&mut self) -> Result<()> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            self.check_role(caller, Role::Owner(this))?;
+            self.env().terminate_contract(caller)
+        }
+
+        /// Result wrapped version of `has_role`
+        ///
+        /// Returns Error variant MissingRole(Role) if the account does not carry a role
+        #[ink(message, selector = 5)]
+        pub fn check_role(&self, account: AccountId, role: Role) -> Result<()> {
+            if !self.has_role(account, role) {
+                return Err(AccessControlError::MissingRole(role));
+            }
+            Ok(())
+        }
+
+        fn emit_event<EE: EmitEvent<Self>>(emitter: EE, event: Event) {
+            emitter.emit_event(event);
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use ink_lang as ink;
+
+        #[ink::test]
+        fn access_control() {
```

Comment:
This is positive case but a unit test for unsuccessful access granting is missing.